### PR TITLE
feat!: refactor action to be CLI based

### LINKED ISSUE #281: Refactor and rename this GH Action to be fully AsyncAPI CLI based
When I worked on this action some time ago, my main motivation was:
- provide GH action for generator
- do it by using Generator as library because:
    - we have automated releases (for node projects) implemented in AsyncAPI organization, so I knew that whenever new Generator is released, it will trigger release of this action to also support latest release
    - because wanted to use Generator as library in production and figure how that works 🤷🏼 

Now times are different 😄 

I think it would be better:
- release 3.0 version of the action
- rename action from `Generator for AsyncAPI documents` to `Generator, validator, converter and others - all in one for your AsyncAPI docs`
- rename repository from `github-action-for-generator` to `github-action-for-asyncapi-cli`
- throw away all the Nodejs code and modify docker file to simply expose the AsyncAPI CLI interface - example GH Action that does it https://github.com/gaurav-nelson/github-action-markdown-link-check

Or:
Or we just archive this repo and create action directly in AsyncAPI CLI - example how a friend does it (CLI and GH action for it in one repo) https://github.com/mszostok/codeowners-validator. I think it makes sense, especially that in ourcase GH Action would anyway always reflect the CLI version. Current action we would just archive, set as deprecated and link to the new one.

Thoughts?

### PR DESCRIPTION
**Pull request by Shurtu-gal**
### Description
The PR changes the GitHub action from being just a generator to an all-in-one tool.

### What's New:
- Introduced new commands: **validate, convert, optimise**.
- Enabled model generation by language.

### Changes Made
- Removed all NodeJS files and dependencies from the GitHub Action.
- Refactored the action to expose the Command-Line Interface (CLI) functionality through the Docker container using a bash script.
- Introduced a Makefile to streamline and simplify the development process.
- Implemented a bash script that dynamically executes commands based on user inputs, transforming the action into a comprehensive interface for CLI operations.

### Breaking Changes:

Previously, the parameter option was exclusive to template parameters. It has now been expanded to apply to all commands. Therefore, users must now add -p before template params for the same functionality.

### Things left

- [x] Update README
- [x] Add tests to test various edge cases
- [x] Write a test workflow for production tests
- [ ] Release 3.0
- [ ] Rename Repository.

**Related issue(s)**
Fixes #281

### CODE REVIEW

**Review by derberg — CHANGES_REQUESTED** [review]
I have some questions about your decisions to do things in a specific way

#### config file

why did you decide to introduce a config file. You executed it perfectly, but is this really needed. In past I faced a similar design decision: should I do a config to allow multiple runs, or should I basically document how to have workflow with multiple steps that use the same action, and I choose to drop idea with config. Config adds a lot of additional maintenance effort, but also makes things harder to use. Also I think that when you look at the ecosystem of GitHub Actions, the general principle is to have a single run responsibility. Example: I have a workflow where I need to checkout 2 different repositories, and to do it, I have 2 separate steps, one to checkout repo A and one to checkout repo B - 2 separate steps, not one step with 2 checkouts. It makes future debugging much much easier, as you can clearly see what step failed.

so what I think is better workflow for any platform/devops team is when they can do:
```yaml
    - name: Validating AsyncAPI document
      uses: docker://asyncapi/github-action-for-cli:3.0.0
      with:
        command: validate
        filepath: ./asyncapi.yml
    - name: Generate models
       uses: docker://asyncapi/github-action-for-cli:3.0.0
       with:
        command: generate
        language: typescript
        filepath: ./asyncapi.yml
    - name: Generating HTML
      uses: docker://asyncapi/github-action-for-cli:3.0.0
      with:
        command: generate
        filepath: ./asyncapi.yml
        template: @asyncapi/html-template@0.9.0
        output: generated-html
```

so you use the same action for 3 different outputs. You know what I mean? this is the best approach from someone running and debugging pipelines. And to configure anything, you only learn inputs and not inputs with config

also, in the past nobody ever requested to do a config approach.

and great stuff is that if we do above, then we will not do any breaking change except for action name change

below can still work. We can say that by default, if `command` is not provided, then we assume `generate` and rest will still work
```
    - name: Generating HTML from my AsyncAPI document
      uses: docker://asyncapi/github-action-for-generator:2.0.0 #always use latest tag as each is pushed to docker
      with:
        template: '@asyncapi/html-template@0.9.0'  #In case of template from npm, because of @ it must be in quotes
        filepath: docs/api/my-asyncapi.yml
        parameters: baseHref=/test-experiment/ sidebarOrganization=byTags #space separated list of key/values
        output: generated-html
```

and yeah, do not introduce input if it is not really needed, like `type` for example, if you can use other inputs, like `template` or `language` to identify if `model` or `fromTemplate` should be run - then do it. Simply validate and throw error when people use `template` and `language` at the same time. Also in docs simply clarify what these mean.

#### CLI version

I think we need to stick to approach that docker contains only one CLI version. With CI/CD pipelines it is all about speed, we cannot make pipeline run longer because CLI installation needs to take place first. So we for sure need to have latest there by default. If you really think `cli_version` is useful, like for testing pipelines with new cli without bumping version in action - then we need a script that basically removes latest and installs requested CLI version - but it also has to be documented, so people understand how that works

---
don't feel disappointed with my feedback - it is only design related, but the execution mate - docs and quality are 💯 ❤️

File: Makefile
Comment (Shurtu-gal) [reply Makefile:8]: This was for testing locally, will remove them.

File: README.md
Comment (Shurtu-gal) [reply README.md:45]: The problem with the shell is that undefined may lead to some errors. I will see if I can do something better.

**Comment by Shurtu-gal** [comment]
I will take out the config approach and keep it simple. Same with cli_version too.

File: action.yml
Comment (derberg) [inline action.yml:36]: what does it represent in CLI? `--debug`?

File: action.yml
Comment (derberg) [inline action.yml:32]: we do not need it right? there should be no custom commands in the end?

File: action.yml
Comment (derberg) [inline action.yml:20]: make sure in the code and in description and docs what fields are mutually exclusive, like that `language` cannot be used with `template` of that in case of `command: validate` other inputs are ignored

just make sure proper logs messages are there, and proper docs in place

File: Makefile
Comment (derberg) [inline Makefile:1]: what will be the strategy to run this test?

till now we had package.json and we used https://github.com/asyncapi/github-action-for-generator/blob/master/.github/workflows/if-nodejs-pr-testing.yml

maybe best would be to leave package.json, even if just there for running tests 🤔 to not have to much CI/CD maintenance on our side. So basically a script in package.json that triggers the make file when `npm run test` is triggered?

wdyt, then we can still use global file https://github.com/asyncapi/github-action-for-generator/blob/master/.github/workflows/if-nodejs-pr-testing.yml

**Review by derberg — COMMENTED** [review]
left few comments, well done 👏🏼

File: action.yml
Comment (Shurtu-gal) [reply action.yml:32]: There are custom commands like if somebody wants to use the commands other than the provided ones.

File: action.yml
Comment (Shurtu-gal) [reply action.yml]: Removed it for now. Can be added later.

File: Makefile
Comment (Shurtu-gal) [reply Makefile]: A quick question:- Do we need js tests? It doesn't make sense when we are using the shell to run everything.

**Comment by Shurtu-gal** [comment]
@derberg README is also updated

File: Makefile
Comment (derberg) [reply Makefile]: we definitely do not need to have js tests

I'm suggesting `package.json` not to have js tests but to be able to reuse global test action that depends on `package.json` - you can still have other tests, make files - just trigger of test will go through package.json

so if you keep package.json and have a `test` script there, we can still use https://github.com/asyncapi/github-action-for-generator/blob/master/.github/workflows/if-nodejs-pr-testing.yml as it triggers test script in https://github.com/asyncapi/github-action-for-generator/blob/master/.github/workflows/if-nodejs-pr-testing.yml#L72

does that make sense?

or maybe we do not need MAKE file, and test should be basically a workflow that represents all possible options and runs on PR

have a look at https://dev.to/cicirello/how-to-test-a-github-action-with-github-actions-2hag 

it works this way that in PR workflow, you do instead of `uses: docker://asyncapi/github-action-for-generator:2.0.0` something like `uses: ./` so like run a workflow with steps that use action version that is in the PR - you know what I mean?

File: package.json
Comment (derberg) [inline package.json:9]: I guess we do not need it, especially that versioning of this action will not me recorded in this file - it is purely development related file. You can also remove `version` property and in `description` specify purpose if that file

you can also remove `package-lock.json` as there are no dependencies to track here

Comment (Shurtu-gal) [reply package.json]: I guess node version bump can also be removed from actions then.

File: CODEOWNER
Comment (derberg) [inline CODEOWNER:9]: @magicmatatjahu pinging you so you are aware

@Shurtu-gal completely rewrites this action and I proposed to him to be maintainer after we merge

are you ok, and also are you still interested to be a maintainer here?

Comment (magicmatatjahu) [reply CODEOWNER:9]: I can be still to accept PRs when they are very important :) @Shurtu-gal welcome!

Comment (Shurtu-gal) [reply CODEOWNER:9]: Thanks

File: Makefile
Comment (derberg) [inline Makefile:16]: can you name them in more transparent way, like `generate-`something ? also add description to each test, like a comment you know that provides human-readable name of test, and best would be what AsyncAPI CLI command is tested - as you know with these default values it is not really readable

File: README.md
Comment (derberg) [inline README.md:25]: I'm still not sure what is the goal of custom command solution 🤔 why cannot we have all commands here?

**Review by derberg — COMMENTED** [review]
left 2 comments

the only thing is missing is a test workflow that runs this action on PR

File: README.md
Comment (Shurtu-gal) [reply README.md:25]: The problem is we would have to make too many switch statement as it is not one format for all commands so I have some basic commands so that the existing ones won't break.

File: Makefile
Comment (Shurtu-gal) [reply Makefile:16]: Okay

**Comment by aeworxet** [comment]
@asyncapi/bounty_team

File: Makefile
Comment (Shurtu-gal) [reply Makefile]: Done

**Comment by Shurtu-gal** [comment]
I have added the test workflow. @derberg 
Sample workflow can be found [here](https://github.com/ash17290/github-action-for-generator/actions/runs/7304146453)

File: .github/workflows/test-action.yml
Comment (derberg) [inline .github/workflows/test-action.yml:27]:
```suggestion
          echo "Listing all files"
          ls -R
          echo "Asserting GitHub Action"
          if [ -f "./output/asyncapi.md" ]; then
            echo "Files exist"
          else
            echo "Files do not exist:- ./output/asyncapi.md"
            echo "Action failed"
            exit 1
          fi
```

please modify all assertions to have it

File: .github/workflows/test-action.yml
Comment (derberg) [inline .github/workflows/test-action.yml:102]: same as with the other question above

File: .github/workflows/test-action.yml
Comment (derberg) [inline .github/workflows/test-action.yml:106]: this workflow should run only on PR, so no need to drop errors to slack as they will be visible on PR

like https://github.com/asyncapi/github-action-for-generator/actions/runs/7304150811/job/20169330511?pr=348

File: Makefile
Comment (derberg) [inline Makefile:1]: do we still needs this make file? isn't it covered with test-action already? or do you think it is needed for local development?

File: .github/workflows/test-action.yml
Comment (derberg) [inline .github/workflows/test-action.yml:90]: you need also one test for `optimize` command

File: .github/workflows/test-action.yml
Comment (derberg) [inline .github/workflows/test-action.yml:89]: please echo some useful comment, that all is good as error was triggered

shouldn't you create here some file? like some `success` file` and then in another step check if file was created, then it means job successfully failed as expected - if it would not fail on `Test GitHub Action` then file would never be created 🤔 and job could fail

**Review by derberg — CHANGES_REQUESTED** [review]
left some comments

can you at least add commands: bundle, diff, and convert that have use case to be used in CI?

File: Makefile
Comment (Shurtu-gal) [reply Makefile:1]: Yep, this is needed for local development. Passing parameters is very tedious especially in bash programs as they have to properly ordered.

File: .github/workflows/test-action.yml
Comment (Shurtu-gal) [reply .github/workflows/test-action.yml:89]: I don't see the point of making a file as on error the outcome is set to failed, but step continues due to `continue-on-error`.


> As seen in docs :- 
> The result of a completed step before continue-on-error is applied. Possible values are success, failure, cancelled, or skipped. When a continue-on-error step fails, the outcome is failure, but the final conclusion is success.

Comment (Shurtu-gal) [reply .github/workflows/test-action.yml:89]: Will add the comment.

**Comment by Shurtu-gal** [comment]
@derberg The requested changes have been added.
Regarding sonarcloud hotspot, I had made the changes to fix it but it led to permission issues when making docker container for GitHub action as can be seen [in the run here](https://github.com/ash17290/github-action-for-generator/actions/runs/7303702561/job/19904669097)

**Comment by derberg** [comment]
@Shurtu-gal please bring back the user settings in dockerfile and simply silent sonar by adding https://github.com/asyncapi/cli/blob/master/.sonarcloud.properties file to this repo, just adjust it to ignore dockerfile, and add comment why we ignore it: that this docker file is for starting containers that should run only in GitHub Actions and no other environments, and running with security violation is ok

**Comment by Shurtu-gal** [comment]
Done @derberg.

**Comment by derberg** [comment]
works well on my side

because we know there is one potential breaking change in case of `parameters` field, and that `-p` needs to be used, I added `!` to the title so we release this new version as new major

**Review by derberg — APPROVED** [review]
well done 👏🏼

**Comment by derberg** [comment]
@allcontributors please add @Shurtu-gal for code,test,doc,ideas,maintainance

**Pull request merged**
